### ISSUE TITLE
History does not work anymore if work space was loaded

### ISSUE THREAD

**Issue by mohamed-barakat** [gap-system/gap#2558]
### Observed behaviour

Saving the GAP history does not work anymore if a workspace was loaded during startup. So there is a difference between the behavior of

`gap` and `gap -L xyz`

Without `-L xyz` the history is saved and with `-L xyz` the history is not saved. In both cases GAPInfo.History starts with the content of the history file and ends with the commands of the current session.

### Expected behaviour

The history should be saved regardless of a workspace being loaded or not.

### Copy and paste GAP banner (to tell us about your setup)
```
 ┌───────┐   GAP 4.9.1 of 05-May-2018
 │  GAP  │   https://www.gap-system.org
 └───────┘   Architecture: x86_64-apple-darwin17.5.0-default64
 Configuration:  gmp 6.1.2, readline
 Loading the library and packages ...
 Packages:   AClib 1.3, Alnuth 3.1.0, AtlasRep 1.5.1, AutPGrp 1.9, Browse 1.8.7, CRISP 1.4.4, Cryst 4.1.17, CrystCat 1.1.8, CTblLib 1.2.2, FactInt 1.6.2, FGA 1.4.0,
             GAPDoc 1.6.1, IO 4.5.1, IRREDSOL 1.4, LAGUNA 3.9.0, Polenta 1.3.8, Polycyclic 2.13.1, PrimGrp 3.3.1, RadiRoot 2.8, ResClasses 4.7.1, SmallGrp 1.3,
             Sophus 1.24, SpinSym 1.5, TomLib 1.2.6, TransGrp 2.0.2, utils 0.54
 Try '??help' for help. See also '?copyright', '?cite' and '?authors'
```

**Comment by ChrisJefferson** [comment]
The deeper bug here is that we don't run any AtExit functions when restoring from a workspace (including cleaning up temporary directories and flushing files)

**Comment by fingolfin** [comment]
Should be fixed by PR #2578, but waiting for confirmation by @mohamed-barakat
